#!/usr/bin/env bash
#PBS -q omp
#PBS -l ncpus=16
#PBS -l mem=40gb
#PBS -l walltime=4:00:00
######################################
#		10/07/2023
#		By Elyna Bouchereau
######################################
echo "#######* MAPPING DATABASES *#######"
#######*	LOAD MODULES		*#####
## Init and load base modules:
#  Load CONDA
. /appli/anaconda/latest/etc/profile.d/conda.sh 
#  And then it's environment
#conda activate bioinfo-ely-env
. /appli/bioinfo/star/2.7.10b/env.sh
echo "Test conda"
conda list | head

## Load packages:
#  and tests the ones needed
PARALLEL="/appli/GParallel/GP-20220722/bin/parallel"
echo "Test packages from conda"
#fastqc --version
star --version
#parallel --version

#######*	DATA SELECTION		*#####
## DATA location
SEQ_RNA_FOLDER="$DATAWORK/Transcriptomic/TRIMMED/"


## SELECTING rRNA DATABASE
export DATABASE="$DATAWORK/Transcriptomic/Genome-Transcriptome"




#######*	PROCESSING PIPELINE	*#####
echo "#######* PROCESSING PIPELINE *#######"
## Generation a folder to put work on
echo "### $DATABASE"

#bowtie2-build -f --threads 8  ${DATABASE} ${DATAWORK}/Transcriptomic/Genome-Transcriptome/Vaestu_transcriptome_DB

#rsem-prepare-reference ${DATABASE} ${DATAWORK}/Transcriptomic/Genome-Transcriptome/Vaestu-transcriptome_DB-RSEM \
#	-p 12 \
#	--bowtie2 

STAR --genomeDir ${DATABASE} \
	--runMode genomeGenerate --runThreadN 16 \
	--genomeFastaFiles ${DATABASE}/VAESTLR12016.1-Contigs.fasta \
	--sjdbGTFfile ${DATABASE}/Vaestu12016.1-Contigs.gtf \
	--sjdbOverhang 74 --genomeSAindexNbases min(14, log2(GenomeLength)/2 - 1) 

echo "### DONE."
